[Edit] Python: __repr__() (#6806)
Codecademy/docs: content/python/concepts/dunder-methods/terms/repr/repr.md

  - 'paths/computer-science'
---

> The **`__repr__()`** method is a special method in Python that returns a string representation of an object. This method is one of Python's dunder (double underscore) methods that provides an "official" string representation of an object. The intention is to create a representation that is unambiguous and, if possible, a valid Python expression that could be used to recreate the object.

Reviewer: ```suggestion
The **`__repr__()`** method is a special method in Python that returns a string representation of an object. This method is one of Python's [dunder](https://www.codecademy.com/resources/docs/python/dunder-methods) (double underscore) methods that provides an "official" string representation of an object. The intention is to create a representation that is unambiguous and, if possible, a valid Python expression that could be used to recreate the object.
```

> The **`__repr__()`** method is a special method in Python that returns a string representation of an object. It is one of Python's dunder (double underscore) methods that provides an "official" string representation of an object, intended to be unambiguous and, if possible, a valid Python expression that could be used to recreate the object.
> 
> The `__repr__()` method is primarily used for debugging and development purposes. When you print an object or display it in an interactive session, Python calls this method to determine what string to show. It's also used when objects are displayed in collections like lists and dictionaries. Implementing a clear and informative `__repr__()` method makes debugging and logging significantly easier, especially for custom objects.

Reviewer: ```suggestion
The **`__repr__()`** method is a special method in Python that returns a string representation of an object. This method is one of Python's dunder (double underscore) methods that provides an "official" string representation of an object. The intention is to create a representation that is unambiguous and, if possible, a valid Python expression that could be used to recreate the object.

The `__repr__()` method is primarily utilized for debugging and development purposes. When printing an object or displaying it in an interactive session, Python calls this method to determine the appropriate string to show. This method is also employed when objects are displayed in collections like lists and dictionaries. Implementing a clear and informative `__repr__()` method enhances debugging and logging significantly, especially for custom objects.
```

## Codebyte Example: Custom Collection Class

This example shows how to implement `__repr__()` for a custom collection class.

```py
class ShoppingCart:
  def __init__(self, owner):
    self.owner = owner
    self.items = {}

  def add_item(self, item, quantity=1):
    self.items[item] = self.items.get(item, 0) + quantity

  def __repr__(self):
    items_repr = ", ".join(f"'{item}': {qty}" for item, qty in self.items.items())
    return f"ShoppingCart('{self.owner}')" + (f".add_multiple({{{items_repr}}})" if self.items else "")

  def add_multiple(self, items_dict):
    for item, qty in items_dict.items():
      self.add_item(item, qty)
    return self  # Allow method chaining

# Create a shopping cart and add items
cart = ShoppingCart("John")
cart.add_item("apple", 5)
cart.add_item("banana", 2)

print(cart)  # Output: ShoppingCart('John').add_multiple({'apple': 5, 'banana': 2})

# This representation can be useful for debugging
empty_cart = ShoppingCart("Alice")
print(empty_cart)  # Output: ShoppingCart('Alice')
```

> This example demonstrates a more complex `__repr__()` implementation for a custom collection class. The representation includes not just the initialization but also a method call to recreate the current state of the object. While this representation isn't as easy to use with `eval()`, it provides a complete picture of the object's state, which is valuable for debugging.

Reviewer: ```suggestion
This example demonstrates a more complex `__repr__()` implementation for a custom collection class. The representation includes not just the initialization but also a method call to recreate the current state of the object. While this representation isn't as easy to use with [`eval()`](https://www.codecademy.com/resources/docs/python/built-in-functions/eval), it provides a complete picture of the object's state, which is valuable for debugging.
```
